### PR TITLE
[WIP] Add categorization to marketplace page

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
Categories are likely to change but that can easily be updated.

https://jira.coreos.com/browse/CONSOLE-764

image

@SamiSousa 
@tlwu2013 
@serenamarie125 

Todo:
- [ ] Refactor Catalog page to use utilities created in this PR
- [ ] Add URL param to allow linking to details of an item

### CODE REVIEW

File: frontend/public/components/marketplace/kubernetes-marketplace-items.jsx [hunk 1]
```diff
@@ -151,62 +208,110 @@ export class MarketplaceTileViewPage extends React.Component {
...
-    }, {});
+    _.forEach(filterGroups, filterGroup => {
+      params.delete(filterGroup);
+    });
+
+    const url = new URL(window.location);
+    const searchParams = `?${params.toString()}${url.hash}`;
+
+    history.replace(`${url.pathname}${searchParams}`);
+    this.setState(this.getUpdatedStateFromURL(searchParams));
```
Comment (dtaylor113) [inline frontend/public/components/marketplace/kubernetes-marketplace-items.jsx]: After clearing filters the keyword textbox should get focus.

Comment (jeff-phillips-18) [reply frontend/public/components/marketplace/kubernetes-marketplace-items.jsx]: Thanks. Fixed.

**Comment by spadgett** [comment]
Is this meant to replace #713? It seems to have some of the same code.

File: frontend/public/components/utils/category-filter-utils.ts [hunk 2]
```diff
@@ -0,0 +1,302 @@
+/* eslint-disable no-undef */
+import * as _ from 'lodash-es';
+
+export class CategoryFilterUtils {
+  static CATEGORY_URL_PARAM = 'category';
+  static KEYWORD_URL_PARAM = 'keyword';
+  static FILTER_SEPARATOR = ',|,';
```
Comment (spadgett) [inline frontend/public/components/utils/category-filter-utils.ts]: I'd rather serialize a JSON array than trying to use our own format

File: frontend/public/components/marketplace/kubernetes-marketplace-items.jsx [hunk 3]
```diff
@@ -225,20 +332,59 @@ export class MarketplaceTileViewPage extends React.Component {
...
+    }
+    if (topLevel && !_.size(subcategories)) {
+      return this.renderCategory(category, true);
+    }
+
+    return (
+      <React.Fragment>
+        {_.map(subcategories, subcategory => this.renderCategory(subcategory, false))}
+      </React.Fragment>
+    );
```
Comment (spadgett) [inline frontend/public/components/marketplace/kubernetes-marketplace-items.jsx]: Can this be

```jsx
return _.map(subcategories, subcategory => this.renderCategory(subcategory, false));
```

**Comment by spadgett** [comment]
Nothing to change in this PR, but we should look at switching to native lazy loading in React 16.6. It might solve some problems for us.

https://reactjs.org/blog/2018/10/23/react-v-16-6.html#reactlazy-code-splitting-with-suspense

@alecmerdler fyi

File: frontend/public/components/utils/category-filter-utils.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
+/* eslint-disable no-undef */
+import * as _ from 'lodash-es';
+
+export class CategoryFilterUtils {
```
Comment (alecmerdler) [inline frontend/public/components/utils/category-filter-utils.ts]: Class with all static methods == module with pure functions.

Comment (jeff-phillips-18) [reply frontend/public/components/utils/category-filter-utils.ts]: but exporting so many functions...

Comment (jeff-phillips-18) [reply frontend/public/components/utils/category-filter-utils.ts]: Fixed.

File: frontend/public/components/utils/category-filter-utils.ts (hunk 2 above)
Comment (jeff-phillips-18) [reply frontend/public/components/utils/category-filter-utils.ts]: Fixed

**Comment by jeff-phillips-18** [comment]
Looking into the use of _.memoize but thus far no luck.

**Comment by spadgett** [comment]
> Looking into the use of _.memoize for lazy route loading, but thus far no luck.

:( ok, no worries

**Comment by jeff-phillips-18** [comment]
Closing in lieu of https://github.com/openshift/console/pull/805

**Pull request closed without merging**